### ISSUE TITLE
Racket repl hangs after closing neovim

### ISSUE THREAD

**Issue by hendersonreed** [Olical/conjure#185]
Hey there, me again :smile: 

## Summary

I've encountered a minor bug I think with the Racket stdio client. After opening a Racket file with neovim, the repl starts as expected. After closing neovim, the shell prompt doesn't appear because the repl process continues in the background.

Hitting ctrl-c brings back the bash prompt, *but the repl session continues in the background, continuing to consume resources.* This occurs each time you open and close a racket file, meaning you could easily get 10 or more different racket repl instances running in the background.

I was able to confirm that it is the repl session that's causing this by manually closing the repl with `<leader>cS` and then quitting. Doing that first means that closing vim works as expected and I end up back at the shell prompt immediately.

### Steps to reproduce: 

1. Open a file with the `.rkt` extension in neovim and verify that the 'started' message appears in the HUD.
2. Close that file with `:x` or :q!`
3. Observe the fact the bash prompt doesn't appear.
4. Hit ctrl-c to exit
5. run `pgrep racket` to see the running racket session that persists.

### Expected behavior (for me at least):

Closing vim should also close the repl session.

**Comment by callum-oakley** [comment]
Ah! This is happening in #187 too, so it's not something peculiar to Racket.

**Comment by Olical** [comment]
Oh interesting! What's weird is that this doesn't happen to me... so it'll be a little more subtle I guess.

**Comment by hendersonreed** [comment]
Your mentioning that it doesn't work in your environment made me go back and check to verify that it wasn't say, my `zsh` setup.

As it is, I was able to replicate it across 2 shells (zsh and bash)  on one install (Fedora 31). However, on a second install with Fedora 33, I wasn't able to replicate it. These machines have roughly similar plugins and configurations. The working installation has nvim 0.4.4 and the installation that demonstrates the bug has nvim 0.4.3.

This is also with both installations running Conjure v4.15.0

So to recap, the general results are:

Bug is present on machine with:
Fedora 31, nvim 0.4.3

Bug is not present on machine with:
Fedora 33, nvim 0.4.4

I've tested it with Conjure 4.15.0, 4.14.0, and 4.8.0 and this issue has been present in all of them.

I'd be curious to hear what kind of environment @callum-oakley is encountering it in, and if they're able to reproduce it with Racket as well as their mit-scheme addition.

**Comment by Olical** [comment]
Haven't had a chance to look yet, sorry :sob: but my hunch is that I won't be able to work out the root cause, but that's okay. I think all clients should listen for Neovim's exit autocmd and shut their sub-processes / connections down before exit. The Clojure client does this, I want to make sure all clients have an easy hook to do this same thing without too much repeated code.

I think I saw something similar in some envs with Clojure before I added that "disconnect on exit" hook. It's just a cleaner shutdown process, rather than hoping Neovim closes things up nicely for us, something libuv might not like.

**Comment by hendersonreed** [comment]
OK yeah, that sounds like a reasonable solution! I was poking at the source to see if that was how it was handled in any of the other client but didn't look at the Clojure client and missed it.

**Comment by Olical** [comment]
I've just added an `on-exit` hook for all clients that gets called on all loaded clients when the `ExitPre` autocmd fires. So hopefully this issue is gone now? If not hopefully it's a large step in the right direction! I'd love some feedback on this :smile:

**Comment by Olical** [comment]
3 seconds!? I wonder what it's doing... I definitely don't see anything like this and never have. I wonder if I need to send harsher signals to the sub processes :thinking: seems weird though.

Also that line is from the guile client... an entirely different lua module which shouldn't be loaded at this point. Unless two clients are being loaded due to them both overlapping with "scheme" :thinking:

This difference might be to do with what filetype your nvim assigns to `*.scm`, `*.rkt` and `*.guile`... I wonder if there's a bug where if two clients match it behaves weirdly, which _could_ be the root cause of the hang on exit. I added some magic so that the `scheme.guile` filetype would match the `scheme` or `guile` clients in Conjure... that might have backfired.

**Comment by Olical** [comment]
I can see the same guile warning in the racket client too! An issue I can reproduce is an issue I can fix. Probably.

**Comment by Olical** [comment]
```
"~/repos/Olical/conjure/dev/racket/other.rkt" 6L, 59C
=== scheme conjure.client.guile.socket
Press ENTER or type command to continue
=== racket conjure.client.racket.stdio
Press ENTER or type command to continue
```

Ah, hah!

**Comment by Olical** [comment]
So my latest change might fix your issue, but I think I want to adjust how this works. The whole dotted filetype thing is interesting, but I think I want to find a way to ignore that feature and instead keep things simple. I want the following to be true:

 * `racket => racket.stdio`
 * `scheme.guile => guile.socket`
 * `scheme => scheme.stdio`

The dotted filetype support I tried to add makes this assignment tricky since `guile` uses a filetype dotted pair and Racket uses another hack entirely, it cycles between the `scheme` and `racket` filetypes to make sure all mappings are loaded!

**Comment by Olical** [comment]
So hopefully I've fixed a fair few things now :sweat_smile: we'll see. I reaaaaaly hope the shutdown is clean now, but I have my doubts, it seems like there's a worse issue lurking there somewhere.

I've basically allowed you to filter Conjure clients down to filetypes and (optionally) file extensions now. This should fix the ambiguity between Racket, MIT Scheme (#187) and Guile.

**Comment by hendersonreed** [comment]
The filetype ambiguity seems to be fully resolved, so that's awesome.

Shutdown unfortunately still seems to be taking a bit of time, and unfortunately I don't know of a good way to get more information out for you.

**Comment by Olical** [comment]
Hmm okay, well there's another attempt on develop now. I'm being far more aggressive with my destruction of the stdio resources.

I still can't reproduce this on Neovim 0.4.4 in NixOS Linux, even when I lock up the sub process with a `(sleep 10)` call, it exits cleanly and immediately. I'd love to be able to reproduce this, just so I could understand it a little. I do wonder if it's just a bug with Neovim 0.4.3 from what you were saying @hendersonreed :thinking:

**Comment by Olical** [comment]
I actually just noticed a racket and mit-scheme process stuck at 100% after my nvim was all shut down. Maybe I have a repro sort of :eyes:

**Comment by hendersonreed** [comment]
Good news! I'm not getting any hanging at all anymore, and not even a slow shutdown - as soon as I quit nvim I am back in the shell with a prompt immediately. So that's all resolved!

There might have been a regression with the filetype detection though? I'm seeing this when opening up any `.rkt` file:
```
Error detected while processing FileType Autocommands for "scheme":
E5105: Error while calling lua chunk: ...local/share/nvim/plugged/conjure//lua/conjure/client.lua:314: No Conjure client for filetype: 'scheme'
Press ENTER or type command to continue
```

`:set filetype?` returns `filetype=racket`, so I'm not sure where it's getting `scheme` from.

**Comment by Olical** [comment]
I'm running into that too! And I have a local change that basically totally fixes the process shutdown (I think!). I think I was just doing it wrong or in a way that doesn't play nice with some processes. I now send a friendly "kill" signal to the process which closes everything down nicely, I think.

I also hooked the mit-scheme client up to the "on-exit" hook so that tries to cleanly shutdown as nvim exits.

I'm looking at the filetype issue, it's really weird how Racket buffers are being set as `filetype=scheme`, I don't remember it being like that, I'm pretty sure it was always `racket`...

All of these various Schemes are causing complications in my understanding of `filetype -> client` mappings! But exciting times to have three different approaches to Scheme floating around, I'd love someone to do a "How to do SICP with Conjure" guide some day.

**Comment by Olical** [comment]
Did this fix this morning and forgot to push, I hope that helps! I need to work out the filetype issue though... no idea why `.rkt` files aren't reporting a `racket` filetype all the time. They should do imo, but maybe I need to find a workaround for it if the ftplugin works in an "interesting" way.

**Comment by hendersonreed** [comment]
With that most recent update, shutdown appears to be reliable (it happens every time) but slowish (2-3 seconds typically.) Using `destroy` was definitely quicker, but I'm not sure what the issues with unclean shutdown might be.

**Comment by taw10** [comment]
I've been following this because of the changes to dotted filetypes, and testing on my setup with Guile.  There seems to be a slight problem (as of 4c0c8af4cf60a6a0).  With this config (which is the default):
```
let g:conjure#filetype#guile = "conjure.client.guile.socket"
```
...opening a file with type `scheme.guile` reports that the MIT Scheme client has been started instead of Guile, but it does appear to actually use the Guile client:
```
 Sponsored by @campbellr ❤                                                                                              
; --------------------------------------------------------------------------------                                       
; "mit-scheme" (started)                                                                                                 
; --------------------------------------------------------------------------------                                       
; /home/taw/starlet/guile.socket (connected)                                                                             
; Empty result                                                                                                           
; --------------------------------------------------------------------------------                                       
; eval (root-form): (+ 1 1)                                                                                              
2                                                                                                                        
```
MIT Scheme is not installed on my system.  I only use Guile, so I can configure for Scheme "globally" to work round the issue.  With the following, it works exactly like expected:
```
let g:conjure#filetype#scheme = "conjure.client.guile.socket"
```
**Possible over-analysis:** It seems to me that `current-client-module-name` works from right to left, but doesn't stop at the first match, so gives higher priority to "scheme" than "guile" with `scheme.guile`.  I would expect it to work the other way round, based on how dotted filetypes are interpreted by Vim itself (rightmost filetype gets the "last word" on settings).

P.S. I can confirm from experience that working through SICP with Conjure is very helpful and a lot of fun!

**Comment by Olical** [comment]
Hmm, that's frustrating. I'm only now realising how messy these scheme filetypes are, they all seems to start as `scheme` then flip to other things and sometimes back.

 * Racket: `scheme -> racket -> scheme` (although I feel like I remember it staying on `racket`? And if you `:e` again it seems to work!)
 * Guile: `scheme -> guile.scheme` (I think?)
 * Scheme: `scheme`

Giving this some thought tonight, hopefully I can come up with something that helps us configure things without overlap, despite filetypes flipping around. Until then, I'd recommend overriding the `g:conjure#filetype#scheme` config, as shown, or `g:conjure#filetypes` to be something like `["clojure", "fennel", "janet", "racket"]` etc.

**Comment by Olical** [comment]
So I've made a small change... I've made the MIT Scheme client the default over Guile since it has an easier "just open a file and go" story. I've documented all of this and added examples of configuration (what you posted, @taw10) to the help and wiki pages.

I've also ensured that Racket and Guile/MITScheme files work separately and don't collide any more. I hope. I think I may still need to make some tweaks around the stopping of child processes, but I'm not sure yet. I still can't reproduce that hang on exit, I wish I could!

I did reproduce processes hanging around at 100% CPU though, so I've atleast fixed that AFAIK, but will try to repro, I'd imagine that issue and the slow exit are related. I'll have a look at this after work tomorrow / over the weekend.

**Comment by hendersonreed** [comment]
You've been incredibly responsive about this issue, so thanks so much! Conjure is really cool, and is probably the best vim+lisp experience I've had. I'm here to answer any questions you have, test new changes, etc.

**Comment by Olical** [comment]
So I think this is the last change I'll make for now. I've fixed a bug that meant my shutdown calls weren't actually firing unless you had the right buffer open, so hopefully this fixes a bunch of edge cases.

I've added as many closes and shutdown hooks as I can, if it's still hanging on exit I don't think I'll be able to help much until I know exactly how to repro it :disappointed: hopefully that helps, definitely fixed some issues.

And now to shuffle the guile and mit clients around into "scheme.stdio" and "scheme.socket"! I realised these are really just general scheme clients at this point and I might as well name them as such + config to change which scheme you launch.

**Comment by hendersonreed** [comment]
Sounds good to me, thanks again!

**Comment by Olical** [comment]
I'm going to close this since I can't reproduce it and I've done all I can for now to alleviate this. I'll get it all released and into the hands of more people, then we can assess if more needs to be done. I read up on what others do to shut down libuv processes and I'm now doing _more_ than some other systems 😅

So at this point I think I'd only be able to reproduce it with certain REPLs and maybe certain OS / hardware setups? If this is still an issue that's affecting a lot of people moving forward I think I'll go down the route of building a VM that replicates someone else's environment that's experiencing it. I'd like to avoid that if possible because _tiiiiiime_ 🙂 but then again, maybe it's fine now!

I definitely fixed some issues and it works flawlessly for me now, but the delay on the shutdown might still persist, I just can't know for now. Thanks for giving me as much info as you did! It really helped!
